### DISCUSSION TITLE
Cumulus config in NCLU style

### DISCUSSION THREAD

**Discussion by ssasso** [ipspace/netlab#48] [Ideas]
I was thinking it could be useful to have the Cumulus config in the Cumulus' "nclu style" (basically a list of "net something" commands - and the related commands to fetch the current config).
I am in doubt if to open a PR to patch the current templates, or to create a new config "model" called cumulus-nclu (or something like that).

Which approach would you prefer?

**Comment by ipspace** [comment]
That's an interesting idea. Considering one could configure a Cumulus device either way (using FRR **vtysh** even after it's been configured with **nclu**) I don't think it makes sense to create a separate device. After all, the initial configuration is done behind the scenes, so as long as it works nobody cares how it's done.

There's just one tiny problem: today I could reuse the FRR config for Cumulus VX (or vice versa). There are two distinct files for every configuration module, but once I have something running on one platform I can mostly copy it over. With the **nclu** approach they would diverge.

As long as we're not adding new functionality to Cumulus VX it doesn't matter, but if I decide to implement new stuff (VRFs and VLANs come to mind), then it might not appear on Cumulus unless someone else does the job -- I'm too old to learn yet another CLI ;)

To recap: if you feel like doing all the heavy lifting to reimplement existing functionality in **nclu** go for it and use existing templates.

Thank you!

**Reply by ssasso** [reply]
And what about using a configuration flag and keep both the options in the current template files?
Something like:
```
{% if defaults.cumulus.use_nclu is defined -%}
# net commands to execute
{%- else -%}
# FRR style configuration
{%- endif %}
```
In that way you can easily support new features starting from the frr based configuration, and that would work flawlessy from day 0 (and in the future also support the NVIDIA's new "nvue" config method simply adding and "elif") - and I can take care of future "porting" of frr config to NCLU, if you want.

thanks
Stefano

**Reply by ipspace** [reply]
I really don't care about what configuration method is used ;) but I don't think it makes sense to use both configuration methods _for the same module_. 

Each VX/FRR template specifies what configuration method it wants to use (if there's `#!/bin/bash` at the top execute the script, otherwise send the script through `vtysh`), so each module could do whatever, and it's perfectly OK to have some modules use nclu and others FRR.

On top of that, if you're volunteering to port future stuff (should it happen) to NCLU then who am I to say "no" ;)) Go for it and submit a PR.

Thank you! Much appreciated!

**Comment by ipspace** [comment]
Implemented a long time ago in cumulus_nvue device (still not sure whether that was a good decision though)
